### ISSUE TITLE
systemd-boot exits on secure boot error w/o trying next boot option

### ISSUE THREAD

**Issue by jdharms** [systemd/systemd#25548]
### systemd version the issue has been seen with

249.11—ubuntu3.4

### Used distribution

Ubuntu 22.04

### Linux kernel version used

5.15.0-43-generic

### CPU architectures issue was seen on

x86_64

### Component

sd-boot

### Expected behaviour you didn't see

I installed systemd-boot on a machine with two unified kernel images, one signed and one not.  I set the unsigned one to be the default boot option.  When booting, I expected sd-boot to attempt to boot the unsigned one and fail and then continue on and try the signed one (which would succeed).

### Unexpected behaviour you saw

When sd-boot attempted to boot the unsigned UKI, it encountered two errors, exited, and my machine went to the next boot option (removable storage in this case).

I believe what is happening is that `image_start()` is returning an error [here](https://github.com/systemd/systemd/blob/v252/src/boot/efi/boot.c#L2748) and then `goto out` is happening when I would expect this case to be handled by continuing through the boot options (or retrying the selected option according to the automatic boot assessment specification).

### Steps to reproduce the problem

Create two unified kernel images, sign one with a key that is registered in the machine's sb db.  Make the unsigned image the default boot option for sd-boot and attempt to boot the machine.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Error loading \EFI\Linux\linux-5.15.0-unsigned.efi: Security Policy Violation

Failed to execute [entry config name]
(\EFI\Linux\linux-5.15.0-unsigned.efi): Security Policy Violation
```

**Comment by poettering** [comment]
Continued discussion from:

https://lists.freedesktop.org/archives/systemd-devel/2022-November/048597.html

**Comment by medhefgo** [comment]
1. You should consider boot counting logic to do this.
2. This  would probably make TPM measurements unpredictable. In fact, I'd argue for forcing a reboot altogether instead of exiting sd-boot to force stable TPM states.

**Comment by medhefgo** [comment]
> Boot counting logic isn't "working" for this case because when systemd-boot fails, the system goes to the next boot option from the EFI manager's point of view.

That's true, but imho better fixed by rebooting (for TPM) anyways. shim also likes to vanish after loading an image.

Btw, you can replicate boot counting with efibootmgr in userspace. You simply leave the old entries and boot order as is and make a new entry that is booted using the bootnext logic. If it fails at any point the firmware will revert to regular boot order. If it succeeds userspace can put it first in the boot order. You can even add boot counting to this, by having (old) userspace keep track of attempts.

**Comment by xnox** [comment]
one could do multiple systemd-boot boot entries, such that it's like systemd-boot; systemd-boot; pxe boot etc.

**Comment by jdharms** [comment]
@medhefgo rebooting in this failure case works well for us.  I've made a very small change to sd-boot that triggers a reboot on a failure from the start_image() call and it does cause the firmware to give sd-boot another shot at things (and boot counting works).

I'm happy to submit a PR with whatever y'all determine the "right" fix would be, and feel free to rename or edit this issue if it should be used to track the issue in a more general manner.

**Comment by 1awesomeJ** [comment]
Hello @jdharms, have you been able to resolve this issue?

@medhefgo, I'm a prospective intern on the project, could I get a few pointers from you on how to start?

Thank you.

**Comment by 1awesomeJ** [comment]
Hello @jdharms,
Hello @medhefgo 
Hello @poettering ,
Hello @xnox.

I'm a prospective Outreachy intern. I'd like to work on this issue. 

I need your guidance please. 


I thought about what @jdharms said about the issue coming from this line:
err= image_start(image, entry)

My naive idea is to implement an array of all available kernel images instead, and loop through.

I'm new to real-world projects in C, especially not something as massive as systemd, but I count on your help anyways.

Thanks guys.

**Comment by jdharms** [comment]
Hi @1awesomeJ , sorry for the delayed response.

We've been working around this issue by patching systemd-boot to reboot the machine when that `err = image_start(....);` line returns an error.  I believe that @medhefgo 's points about rebooting being better for TPM unsealing and shim availability are good.  I don't know enough to say what a proper solution to this issue would be or how it would look, but it would be excellent to see one!

**Comment by 1awesomeJ** [comment]
Thank you so much @jdharms.
You assume I'll understand what you wrote though. Well I understand like 30%.

I'm a prospective Outreachy intern who has just joined the community.

I got a lot of help yesterday, and even a merge on my PR eventually.

I hope to grow and make more meaningful contributions soon.

I'll be sure to reach out if I get stuck.

Thank you sir .
